### PR TITLE
Filter non-finite losses

### PR DESCRIPTION
**Pull request by csukuangfj**
See
https://github.com/k2-fsa/icefall/pull/396#issuecomment-1207277213

> The pruned loss is inf. @csukuangfj perhaps we should modify the code that sums up the losses, in icefall/icefall/, to filter out inf's and nan's before incorporating them into the sum?

---

Would be nice if someone can help to propagate the changes to other directories.

### CODE REVIEW

File: egs/librispeech/ASR/pruned_transducer_stateless5/train.py [hunk 1]
```diff
@@ -675,6 +691,7 @@ def compute_loss(
     info = MetricsTracker()
     with warnings.catch_warnings():
         warnings.simplefilter("ignore")
+        # info["frames"] is an approximate number
```
Comment (danpovey) [inline egs/librispeech/ASR/pruned_transducer_stateless5/train.py:694]: perhaps should say why this is so? (presumably due to filtering inf/nan losses).
also, I suggest that if all of the either simple or pruned losses are infinite and the
batch has more than, say, 10 items, we could crash?  it's a waste of compute to 
continue to train a model that has diverged.  (The idea here is that 1 or 2 losses can
be infinite for valid reasons, but it's unlikely for a whole batch of 10; I'm picking a number
such that most setups will sometimes processa at least that many batches.  And it doesn't
require adding extra numbers to keep track off like the number of failures).

Comment (csukuangfj) [reply egs/librispeech/ASR/pruned_transducer_stateless5/train.py:694]: Yes, will add it later today.

Comment (csukuangfj) [reply egs/librispeech/ASR/pruned_transducer_stateless5/train.py]: Fixed. Please have a review again.

Sorry for the late reply.

**Comment by csukuangfj** [comment]
@marcoyang1998 

Could you please help to propagate changes from this PR to other `pruned_transducer_statelessX` folders?

File: egs/librispeech/ASR/pruned_transducer_stateless5/train.py [hunk 2]
```diff
@@ -655,7 +655,35 @@ def compute_loss(
...
+        pruned_loss_is_finite = torch.isfinite(pruned_loss)
+        is_finite = simple_loss_is_finite & pruned_loss_is_finite
+        if not torch.all(is_finite):
+            logging.info(
+                "Not all losses are finite!\n"
+                f"simple_loss: {simple_loss}\n"
+                f"pruned_loss: {pruned_loss}"
+            )
+            display_and_save_batch(batch, params=params, sp=sp)
+            simple_loss = simple_loss[simple_loss_is_finite]
```
Comment (desh2608) [inline egs/librispeech/ASR/pruned_transducer_stateless5/train.py:670]: Just a heads up: I was facing the same logical indexing issue here that we had [seen earlier](https://github.com/k2-fsa/icefall/issues/240#issuecomment-1065402350), so I had to replace this with the `torch.where()`. I suppose this might just be a CUDA 11.1 issue after all.
